keps/sig-storage/1495-volume-populators/README.md

```diff
@@ -271,6 +294,153 @@ downgrade.
 
 No issues
 
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [X] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: AnyVolumeDataSource
+    - Components depending on the feature gate: kube-apiserver
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+  Before this feature, kube API server would silently drop any PVC data source
+  that it didn't recognize, as if the client never populated the field. After
+  enabling this feature, the API server allows any object to be specified.
+  An external admission controller will take over validation of the field,
+  and its response to invalid data sources will be to fail the create operation,
+  rather than silently drop them.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Turning off the feature gate should drop all the data sources that were not
+  already valid and restore the old behavior.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  PVC which previously had data sources that were invalided by the rollback
+  would be empty volumes. Users would need to delete and re-create those PVCs.
+
+* **Are there any tests for feature enablement/disablement?**
+  Not at this time.
+
+### Rollout, Upgrade and Rollback Planning
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  As long as the validating webhook and its associated CRD is installed before
+  enabling the feature gate, the desired result will be achieved. Enabling the
+  feature gate before installing the validation hook will allow some PVCs to
+  be created with invalid data sources, and those PVCs will never get dynamically
+  provisioned.
+
+* **What specific metrics should inform a rollback?**
+  None.
```

Comment:
This surprises me.  It seems like knowing that things like failure rates for the PVC create/update would be valuable.  Also, are there recommended metrics for data sources?

Reply:
So I'm not at all familiar with what metrics exist already around PVCs and volume creation. I was thinking metrics would already exist in this area and new ones would not be needed, but that might just be my ignorance. I need to familiarize myself with existing metrics and think about what new ones would make sense in light of this feature. After that, it should be obvious if any of the metrics would inform a rollback.

Reply:
There's a metric `csi_sidecar_operations_seconds`,  Container Storage Interface operation duration with gRPC error code status total, for all CSI sidecars.  If the controller is not implemented as a sidecar, then this existing metric mechanism won't be available for it.

Reply:
Do we have a metric for a histogram of how long a PVC has been waiting for a PV to bind?  That might be interesting.  Perhaps broken down by provisioner, but beware cardinality

Reply:
The csi_sidecar_operations_seconds metric measure how long the createVolume call took, but it doesn't break it down by type, ie normal, snapshot, clone. I think it's valuable to enhance it to be able to distinguish.

For total e2e time it took to bind, we have an experimental volume_operations_total_seconds metric.